specification/snapshot/_types/SnapshotRepository.ts

```diff
@@ -17,22 +17,126 @@
  * under the License.
  */
 
-import { Uuid } from '@_types/common'
+import { ByteSize, Uuid } from '@_types/common'
 import { integer } from '@_types/Numeric'
+import { Duration } from '@_types/Time'
 
-export class Repository {
-  type: string
+export type Repository =
+  | AzureRepository
+  | GcsRepository
+  | S3Repository
+  | SharedFileSystemRepository
+  | ReadOnlyUrlRepository
+  | SourceOnlyRepository
+
+export class AzureRepository {
+  type: 'azure'
+  uuid?: Uuid
+  settings: AzureRepositorySettings
+}
+
+export class GcsRepository {
+  type: 'gcs'
   uuid?: Uuid
-  settings: RepositorySettings
+  settings: GcsRepositorySettings
 }
 
-export class RepositorySettings {
-  chunk_size?: string
-  compress?: string | boolean
-  concurrent_streams?: string | integer
+export class S3Repository {
+  type: 's3'
+  uuid?: Uuid
+  settings: S3RepositorySettings
+}
+
+export class SharedFileSystemRepository {
+  type: 'fs'
+  uuid?: Uuid
+  settings: SharedFileSystemRepositorySettings
+}
+
+export class ReadOnlyUrlRepository {
+  type: 'url'
+  uuid?: Uuid
+  settings: ReadOnlyUrlRepositorySettings
+}
+
+export class SourceOnlyRepository {
+  type: 'source'
+  uuid?: Uuid
+  settings: SourceOnlyRepositorySettings
+}
+
+export type RepositorySettings =
```

Comment:
Seems like we don't need this type.

Reply:
It gets imported and used [here](https://github.com/elastic/elasticsearch-specification/blob/b3a1e6f30b6a0e19b644c406331e1d56bd2907a0/specification/snapshot/create_repository/SnapshotCreateRepositoryRequest.ts#L46). I'm not sure why the request `body` accepts both `repository` and `settings` keys when `repository.settings` also exists. :shrug:

Reply:
Hmm... I'll look at the ES source code to understand if this is needed. Now `@variants internal tag='type'` won't work here since the union members don't have a `type` property.

Reply:
Good catch on the `type` property. Fixed that.

> Hmm... I'll look at the ES source code to understand if this is needed.

I am still not great at navigating the Java codebase, but if [this is the request](https://github.com/elastic/elasticsearch/blob/1ca66bde9109bcc716a692be4a3531c0418299e8/server/src/main/java/org/elasticsearch/action/admin/cluster/repositories/put/PutRepositoryRequest.java#L34-L40), it certainly does not seem like it supports an optional `repository` property. I scrolled through a lot of the repository creation tests, too, and didn't see it. Feel free to correct my reading of it if I'm totally in the wrong place.